docs/howtos/integrations/unit-testing.md

```diff
@@ -36,11 +36,132 @@ if __name__ == "__builtin__" or __name__ == "builtins":
+        return bool(string)
+    return string
+```
+A naive unit test will be as follows:
+```
+def test_convert_string_to_type():
+    from File import convert_string_to_type
+    string = 'true'
+    assert convert_string_to_type(string) == True
+    
+    string = '432'
+    assert convert_string_to_type(string) == 432
+
+    string = 'str'
+    assert convert_string_to_type(string) == 'str'
+```
+The correct way to test this function is using the @pytest.mark.parametrize fixture:
+```
+@pytest.mark.parametrize('string, output', [('true', True), ('432', 432), ('str', 'str')])
+def test_convert_string_to_type(string, output):
+    assert convert_string_to_type(string) == output
+```
+We declare the inputs and outputs in the following format: 'input, output', [(case1_input, case1_output), (case2_input, case2_output), ...]
+(Note that more than two variables can be delivered)
+
+After declaring the variables and assigning their values, you need to assign the variables to the test function. In the example above we assign the variables 'string' and 'output' to the test function.
+
+To read more on parametrize fixture, visit: https://docs.pytest.org/en/latest/parametrize.html
+
+### Complex objects testing
+Sometimes we need to test functions that contain complex object with multiple attributes. Different edge cases may differ in different attributes values of the complex object.
+Initializing the object several times during the test can be slow and impact performance. 
+Let's examine the following object:
+```
+Class ComplexObject:
+    def __init__(self, b, c, d, e, f):
+        self.b = b
+        self.c = c
+        self.d = d
+        self.e = e
+        self.f = f
+    
+    def method_b(self):
+        if isinstance(self.b, list) and len(self.b):
+            return self.b[0]
+        elif isinstance(self.b, str) and self.b:
+            return self.b.lower()
+        else
+            return self.b
+    
+    def method_c(self):
+        pass
+
+    .
+    .
+    .
+```
+A naive approach to test method_b would be:
+```
+def test_method_b():
+    from File import ComplexObject
+    complex_object = ComplexObject([1], None, None, None, None)
+    assert complex_object.method_b() == 1
+
+    complex_object = ComplexObject('B', None, None, None, None)
+    assert complex_object.method_b() == 'b'
+
+    complex_object = ComplexObject({}, None, None, None, None)
+    assert not complex_object.method_b()
+```
+The correct way to test method_b is patching the ComplexObject object. First, in your test file import the patch method using the following line of code:
+```
+from mock import patch
+```
+Next we can mock the __init__ method of ComplexObject once and then test all cases with parametrize, as follows:
+```
+@pytest.mark.parametrize('b, output', [([1], 1), ('B', 'b'), ({}, True)])
+def test_method_b(b, output):
+    with patch.object(ComplexObject, '__init__', lambda a, b, c, d, e, f: None):
```

Comment:
Is `a` supposed to be `self`? if so let's change it to self

Reply:
Nope, should be random variable name, we are mocking the init function.